homeassistant/components/cover/ryobi_gdo.py

```diff
@@ -0,0 +1,285 @@
+"""
+Ryobi platform for the cover component.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/ryobiGDO/
+"""
+import logging
+import json
+import socket
+from datetime import timedelta
+import voluptuous as vol
+import requests
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.cover import (
+    CoverDevice, PLATFORM_SCHEMA, SUPPORT_OPEN, SUPPORT_CLOSE)
+from homeassistant.helpers.event import track_utc_time_change
+from homeassistant.const import (
+    CONF_DEVICE, CONF_USERNAME, CONF_PASSWORD,
+    STATE_UNKNOWN, STATE_CLOSED, STATE_OPEN, CONF_COVERS)
+
+REQUIREMENTS = ['websocket-client==0.37.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+COVER_SCHEMA = vol.Schema({
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_COVERS): vol.Schema({cv.slug: COVER_SCHEMA}),
+})
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+RYOBI_API_KEY_URL = "https://tti.tiwiconnect.com/api/login"
+RYOBI_DEVICES_URL = "https://tti.tiwiconnect.com/api/devices"
+RYOBI_WS_URL = "wss://tti.tiwiconnect.com/api/wsrpc"
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Ryobi covers."""
+    covers = []
+    devices = config.get(CONF_COVERS)
+
+    for device_id, device_config in devices.items():
+        api_key = retrieve_api_key(device_config.get(CONF_USERNAME),
+                                   device_config.get(CONF_PASSWORD),
+                                   device_config.get(CONF_DEVICE, device_id))
```

Comment:
It doesn't make sense to continue if the credentials are wrong or the device ID.